lib/fastifySession.js

```diff
@@ -128,7 +128,16 @@ function onRequest (options) {
 function onSend (options) {
   return function saveSession (request, reply, payload, done) {
     const session = request.session
-    if (!session || !session.sessionId || !shouldSaveSession(request, options.cookie, options.saveUninitialized)) {
+    if (!session || !session.sessionId) {
+      done()
+      return
+    }
+
+    if (!shouldSaveSession(request, options.cookie, options.saveUninitialized)) {
+      // if a session cookie is set, but has a different ID, clear it
+      if (request.cookies[options.cookieName] && request.cookies[options.cookieName] !== session.encryptedSessionId) {
+        reply.clearCookie(options.cookieName)
```

Comment:
I think a better approach would be to do the `clearCookie` call in https://github.com/fastify/session/blob/fbb54079a9b8609bcb141527711109b616f8ef48/lib/fastifySession.js#L62-L65 instead, but we don't have access to `reply` in there, and `decryptSession` is public API